File: apps/csharp/src/Ch08/BankAccount.cs
using System.Runtime.CompilerServices;

namespace GrokkingConcurrency.CSharp.Ch08;

/// <summary>
/// 銀行口座 - 同期と排他制御の例
/// Lock を使ったスレッドセーフな実装
/// </summary>
public class BankAccount
{
    private readonly object _lock = new();
    private int _balance;

    public BankAccount(int initialBalance)
    {
        _balance = initialBalance;
    }

    /// <summary>
    /// 残高
    /// </summary>
    public int Balance
    {
        get
        {
            lock (_lock)
            {
                return _balance;
            }
        }
    }

    /// <summary>
    /// 入金
    /// </summary>
    public void Deposit(int amount)
    {
        lock (_lock)
        {
            _balance += amount;
        }
    }

    /// <summary>
    /// 出金
    /// </summary>
    public bool Withdraw(int amount)
    {
        lock (_lock)
        {
            if (_balance >= amount)
            {
                _balance -= amount;
                return true;
            }
            return false;
        }
    }

    /// <summary>
    /// 口座間送金（デッドロック回避付き）
    /// </summary>
    public static bool Transfer(BankAccount from, BankAccount to, int amount)
    {
        // デッドロック回避のため、常に同じ順序でロックを取得
        var first = RuntimeHelpers.GetHashCode(from) < RuntimeHelpers.GetHashCode(to) ? from : to;
        var second = first == from ? to : from;

        lock (first._lock)
        {
            lock (second._lock)
            {
                if (from._balance >= amount)
                {
                    from._balance -= amount;
                    to._balance += amount;
                    return true;
                }
                return false;
            }
        }
    }
}

File: apps/fsharp/src/Ch08/BankAccount.fs
namespace GrokkingConcurrency.Ch08

open System.Runtime.CompilerServices

type BankAccount = {
    Lock: obj
    mutable Balance: int
}

module BankAccount =
    /// Create a new bank account
    let createAccount (initialBalance: int) : BankAccount =
        { Lock = obj(); Balance = initialBalance }

    /// Get the current balance
    let getBalance (account: BankAccount) : int =
        lock account.Lock (fun () -> account.Balance)

    /// Deposit money into the account
    let deposit (account: BankAccount) (amount: int) : unit =
        lock account.Lock (fun () ->
            account.Balance <- account.Balance + amount
        )

    /// Withdraw money from the account
    let withdraw (account: BankAccount) (amount: int) : bool =
        lock account.Lock (fun () ->
            if account.Balance >= amount then
                account.Balance <- account.Balance - amount
                true
            else
                false
        )

    /// Transfer money between accounts atomically, avoiding deadlock
    let transfer (from: BankAccount) (toAccount: BankAccount) (amount: int) : bool =
        // Always lock in consistent order to avoid deadlock
        let fromHash = RuntimeHelpers.GetHashCode(from)
        let toHash = RuntimeHelpers.GetHashCode(toAccount)
        let (first, second) =
            if fromHash < toHash then (from, toAccount)
            else (toAccount, from)

        lock first.Lock (fun () ->
            lock second.Lock (fun () ->
                if from.Balance >= amount then
                    from.Balance <- from.Balance - amount
                    toAccount.Balance <- toAccount.Balance + amount
                    true
                else
                    false
            )
        )
